### ISSUE TITLE
Missing 'genhtml_line_med_limit' and 'genhtml_line_hi_limit'

### ISSUE THREAD

**Issue by jpmh1309** [linux-test-project/lcov#162]
When generating the .html I'd like to specify a different threshold for line, branch, and function coverage. For branch and function coverage I'm using: 

```
genhtml_function_med_limit=MED_VALUE
genhtml_function_hi_limit=HI_VALUE
genhtml_branch_med_limit=MED_VALUE
genhtml_branch_hi_limit=HI_VALUE
```

But, for line coverage, I couldn't specify the threshold. As a workaround, I'm using:

```
genhtml_hi_limit=HI_VALUE
genhtml_med_limit=MED_VALUE
```

However that also modifies the branch and function threshold, so I had to reset those thresholds to the previous values. So, I'm proposing to introduce: 

```
genhtml_line_med_limit=MED_VALUE
genhtml_line_hi_limit=HI_VALUE
```

**Comment by henry2cox** [comment]
Are you interested only in colorization?  Or are you trying to do some more automated coverage criteria check?

If the latter:  then you can try the ‘--criteria-script callback.sh’ approach that I had implemented (though not on the lcov master branch yet).
We use that in our Jenkins jobs – so Jenkins will mark the job ‘unstable’ if our criteria is not met.
The callback approach enables you to create as simple or as complicated metric as you like.

Implementation-wise:  genhtml calls your callback with a json string of the coverage data.  You return zero or nonzero exit status and possibly stdout/stderr messages.
genhtml returns non-zero exit status if any of your callbacks returned non-zero status.

Henry

**Comment by henry2cox** [comment]
Ach.
I should read the code before replying…

It appears there is already a way to change the colorization thresholds for different types of coverpoints – but you can set the values only through the lcovrc file (not the command line).
See ‘man lcovrc’ for details of parameter name and usage.

Henry

**Comment by jpmh1309** [comment]
Hi @henry2cox ,

But I was looking at the code, it look that it does exist a specific threshold for function coverage (https://github.com/linux-test-project/lcov/blob/master/bin/genhtml#L95) and branch coverage (https://github.com/linux-test-project/lcov/blob/master/bin/genhtml#L99). 

But, for line coverage it uses the global coverage https://github.com/linux-test-project/lcov/blob/master/bin/genhtml#L91. I though, it should only uses the glolabl it is not defined a specific threshold, as happens in branch and function here https://github.com/linux-test-project/lcov/blob/master/bin/genhtml#L398

**Comment by jpmh1309** [comment]
As an example, I'm using this lcovrc file to define the new thresholds of line coverage. As you see, I had to restore the thresholds of branch and coverage.

```
genhtml_hi_limit=15
genhtml_med_limit=30
genhtml_function_med_limit=10
genhtml_function_hi_limit=20
genhtml_branch_med_limit=5
genhtml_branch_hi_limit=10
```
It'd be easier to only call:
```
genhtml_line_med_limit=15
genhtml_line_hi_limit=30
```

**Comment by henry2cox** [comment]
Hi –

I didn’t look at the html generation code too closely – so it is possible that we actually ignore some of the threshold variables – but what I find are 3 sets of thresholds:

# Specify coverage rate limits (in %) for classifying file entries

# HI:   $hi_limit <= rate <= 100          graph color: green

# MED: $med_limit <= rate <  $hi_limit    graph color: orange

# LO:          0  <= rate <  $med_limit   graph color: red



# For line coverage/all coverage types if not specified

our $hi_limit = 90;

our $med_limit = 75;



# For function coverage

our $fn_hi_limit;

our $fn_med_limit;



# For branch coverage

our $br_hi_limit;

our $br_med_limit;

Which are subsequently set after reading the lcovrc file:

# Copy related values if not specified

$fn_hi_limit    = $hi_limit if (!defined($fn_hi_limit));

$fn_med_limit   = $med_limit if (!defined($fn_med_limit));

$br_hi_limit    = $hi_limit if (!defined($br_hi_limit));

$br_med_limit   = $med_limit if (!defined($br_med_limit));

$func_coverage  = $lcov_func_coverage if (!defined($func_coverage));

$br_coverage    = $lcov_branch_coverage if (!defined($br_coverage));

So – as you point out:  function, branch, and line all get the same threshold if you override nothing, or if you don’t explicitly override either branch or function thresholds.
But you _can_ override all of them separately, if you like.

Henry

**Comment by jpmh1309** [comment]
Look at this PR https://github.com/linux-test-project/lcov/pull/163/files, this is what I mean

**Comment by jpmh1309** [comment]
Back to you @henry2cox

**Comment by henry2cox** [comment]
Ah.  Sorry.  Was not thinking of myself as gatekeeper 😊

As I understand it:  this patch enables the user to set branch/line/function colorization thresholds independently – by adding a feature to set the line coverage thresholds without using a master key.
Without the patch:  if you want to set only the ‘line coverage’ thresholds without changing the branch or function thresholds, then you need to specify both branch and line because the ‘master’ values that are used to set line thresholds also set the others by default.
Thus:  this gives you an arguably simpler way to configure your HTML (as opposed to a way to do something new that as not possible previously).

Personal experience is that most projects don’t use branch coverage, and set a much higher function coverage target than line coverage target (and higher line coverage than branch coverage, if they use branch coverage) – so I’m not sure that too many people will use this new flexibility.
On the other hand:  it doesn’t break any existing scripts or use cases – and I have definitely implemented a lot of features for which I am the sole user – so who am I to say? 😊

Peter is the actual arbiter for all things lcov – and I know that he is away/on vacation for the next couple of weeks.

Hope this helped

Henry

**Comment by henry2cox** [comment]
Just merged the associated PR - sha 09704264f0fac.
